paz/backend/boxes.py

```diff
@@ -4,377 +4,552 @@
+    return (index + 1, scores, indices)
+
+
+def create_step_function(x_minimum, y_minimum, x_maximum, y_maximum, areas, IoU_thresh):
+    """Create step function for NMS loop."""
 
     def step(state):
-        i, scores, indices = state
-        best_idx = jp.argmax(scores).astype(jp.int32)
-        updated_indices = indices.at[i].set(best_idx)
-        iou = calculate_IoU_with_best_box(
-            x_min, x_max, y_min, y_max, best_idx, areas
-        )
-        updated_scores = _suppress_overlapping_boxes(
-            best_idx, iou, scores, iou_thresh
+        index, scores, indices = state
+        return apply_NMS_iteration(
+            x_minimum, y_minimum, x_maximum, y_maximum, areas, IoU_thresh, scores, index, indices
         )
 
-        return (i + 1, updated_scores, updated_indices)
-
     return step
 
 
-def _nms_continuation_condition(top_k):
-    """Determines whether to continue the NMS iterations."""
+def handle_empty_case(boxes, scores):
+    """Handle cases with no boxes."""
+    if boxes.shape[0] == 0:
+        return jp.zeros(0, dtype=jp.int32), 0
+    return None
 
-    def condition(state):
-        i, curr_scores, _ = state
-        continue_iteration = jp.logical_and(
-            i < top_k, jp.max(curr_scores) > -jp.inf
-        )
-        return continue_iteration
 
-    return condition
+def compute_geometric_features(boxes):
+    """Compute geometric features (x_min, etc.) from boxes."""
+    x_minimum = boxes[:, 0]
+    y_minimum = boxes[:, 1]
+    x_maximum = boxes[:, 2]
+    y_maximum = boxes[:, 3]
+    areas = (x_maximum - x_minimum) * (y_maximum - y_minimum)
+    return x_minimum, y_minimum, x_maximum, y_maximum, areas
 
 
-def _compute_box_geometric_features(boxes):
-    """
-    Computes geometric features of boxes: x_min, y_min, x_max, y_max, areas.
-    """
-    x_min, y_min = boxes[:, 0], boxes[:, 1]
-    x_max, y_max = boxes[:, 2], boxes[:, 3]
-    areas = (x_max - x_min) * (y_max - y_min)
-    return x_min, y_min, x_max, y_max, areas
+def initialize_NMS_state(scores, top_k):
+    """Initialize state for NMS processing."""
+    selected_indices = jp.zeros(top_k, dtype=jp.int32)
+    return (0, scores, selected_indices)
 
 
-def apply_non_max_suppression(boxes, scores, iou_thresh=0.45, top_k=200):
-    """
-    Applies Non-Maximum Suppression using fixed-iteration loop.
+def create_condition_function(top_k):
+    """Create condition function for NMS loop."""
+    return lambda state: (state[0] < top_k) & (state[1].max() != -jp.inf)
+
+
+def run_NMS_loop(condition_fn, step_fn, init_state):
+    """Run the NMS loop using lax.while_loop."""
+    return lax.while_loop(condition_fn, step_fn, init_state)
+
+
+def process_final_results(final_state, top_k):
+    """Process final NMS results to get selected indices."""
+    count, _, indices = final_state
+    num_selected = jp.minimum(count, top_k)
+    selected = lax.dynamic_slice(indices, (0,), (num_selected,))
```

Comment:
I don't know if this function is jittable? Would it be possible to confirm this, maybe a unit test?